habitat_baselines/il/data/nav_data.py

```diff
@@ -0,0 +1,528 @@
+
+        for i in range(len(actions)):
+            if actions[i] != prev_action:
+                planner_actions.append(actions[i])
+                pq_idx.append(i)
+
+            if i > 0:
+                ph_idx.append(ph_trck)
+                if actions[i] == prev_action:
+                    controller_actions.append(1)
+                    controller_action_ctr += 1
+                else:
+                    controller_actions.append(0)
+                    controller_action_ctr = 0
+                    ph_trck += 1
+                cq_idx.append(i)
+
+            prev_action = actions[i]
+
+            if controller_action_ctr == controller_action_lim - 1:
+                prev_action = False
+
+        return planner_actions, controller_actions, pq_idx, cq_idx, ph_idx
+
+    def get_hierarchical_features_till_spawn(
+        self,
+        idx: int,
+        actions: np.ndarray,
+        backtrack_steps: int = 0,
+        max_controller_actions: int = 5,
+    ):
+        action_length = len(actions)
+
+        pa, ca, pq_idx, cq_idx, ph_idx = self.flat_to_hierarchical_actions(
+            actions=actions, controller_action_lim=max_controller_actions
+        )
+
+        # count how many actions of same type have been encountered before
+        # starting navigation
+
+        backtrack_controller_steps = actions[
+            0 : action_length - backtrack_steps + 1 :  # noqa: E203
+        ][::-1]
+        counter = 0
+
+        if len(backtrack_controller_steps) > 0:
+            while (counter <= self.max_controller_actions) and (
+                counter < len(backtrack_controller_steps)
+                and (
+                    backtrack_controller_steps[counter]
+                    == backtrack_controller_steps[0]
+                )
+            ):
+                counter += 1
+
+        target_pos_idx = action_length - backtrack_steps
+
+        controller_step = True
+        if target_pos_idx in pq_idx:
+            controller_step = False
+
+        pq_idx_pruned = [v for v in pq_idx if v <= target_pos_idx]
+        pa_pruned = pa[: len(pq_idx_pruned) + 1]
+
+        raw_img_feats = self.cnn(self.frame_queue).data.cpu().numpy().copy()
+
+        controller_img_feat = torch.from_numpy(
+            raw_img_feats[target_pos_idx].copy()
+        )
+        controller_action_in = pa_pruned[-1] - 2
+
+        planner_img_feats = torch.from_numpy(
+            raw_img_feats[pq_idx_pruned].copy()
+        )
+        planner_actions_in = torch.from_numpy(np.array(pa_pruned[:-1]) - 1)
+
+        init_pos = self.episodes[idx].shortest_paths[0][target_pos_idx]
+
+        return (
+            planner_actions_in,
+            planner_img_feats,
+            controller_step,
+            controller_action_in,
+            controller_img_feat,
+            init_pos,
+            counter,
+        )
+
+    def preprocess_actions(self) -> None:
+        """
+        actions before -
+        0 - FWD; 1 - LEFT; 2 - RIGHT; 3 - STOP;
```

Comment:
We've been using ` 0 - STOP; 1 - FWD; 2 - LEFT; 3 - RIGHT` in habitat recently.  Probably best to use that here for consistency.